### DISCUSSION TITLE
surgical tool segmentation on live feed, using sam2 and yolo11

### DISCUSSION THREAD

**Discussion by robotricks16** [ultralytics/ultralytics#26125] [Q&A]
hi community,
currently i am working on a project of real surgical tool segmentation on live feed, and after careful research, i came to these conclusions:
**my aim for now** - just detection and segmentation of surgical tools, no classifcation for now

dataset i have used -
positive training - choleseg, rarp, endovis2017,endovis2018, coco random daily life images with instruments mask (let say it Dataset coco_surg, for further reference in the discussion) for background agnostic.
negative training - custom dataset of non surgical instruments like pen, screwdrive highlighter and coco dataset for negative training.

I decided to use the **surgslot sam2 hiera b+ checkpoint** for precise segmentation, but found that , we need to prompt the sam2 with a bbox, or a mask.

**so for autonomous segmentation i decided to train the yolo11s and yolo11s-seg models for autonomously prompting either bbox, mask, or center pca points, but this is where i am stuck from last 1 week,**

**first i trained the yolo11s** 
**issues** - it was prompting the tool mask or bbox to sam2, but also prompting non surgical instruments, which looks like the surgical instruments , like scissors, cylindrical steel rod, etc, even elongated objects like highlighter , pen also 

then i trained the yolo11s-seg model
this time, i actually negatively trained the model using the custom dataset of non surgical instrument and coco everyday things dataset
**issues -  low confidence** - i am not able to achieve high conf on real surgical instruments, sam2 trigger point has been set to 40% conf, but i want to achieve atleast 70-80% conf on surgical tools.

then i ran only yolo11s-seg model trained on this approach of pos and neg dataset, on live feed, turns out that on one instrument i was getting around 50% confidence, and on other i was getting around 20%.

few issues that i noticed earlier:
1. domain shift - majorly data of surgical isntrument have red/pick background due to tissues, so yolo significalty fails on other background thats why i positively train it on coco_surg.

recommendations and help will be appreciated for boosting the yolo conf

**Comment by UltralyticsAssistant** [comment]
👋 Hello @robotricks16, thank you for the detailed surgical-tool segmentation report. The false positives on visually similar non-surgical objects and low YOLO11 confidence on live surgical footage are important issues to investigate, especially under the background domain shift you described.

Please share a minimal reproducible training/inference setup with the exact Ultralytics version, model configuration, class definitions, image and label counts for each dataset split, training arguments, validation metrics, and a few representative validation/live-feed results. This will help distinguish dataset generalization, annotation, and confidence-thresholding issues. Also note that a fixed 70–80% confidence target is not guaranteed; the threshold should be selected from validation results for the desired precision/recall trade-off.

For guidance, please review the [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) and [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips). Upgrade to the latest `ultralytics` package and verify the behavior with `pip install -U ultralytics` before comparing results.

This is an automated response, and an Ultralytics engineer will assist you soon.

**Reply by robotricks16** [reply]
**Environment**

ultralytics==8.4.125
torch==2.13.0+cu130
python 3.10.20
CUDA available, NVIDIA RTX A2000 (12GB)

**Task / model**

task: segment
model: yolo11s-seg.pt (COCO-pretrained checkpoint, fresh start, not resumed)

**Classes**

names:
  0: surgical_instrument
Single class, agnostic across instrument types (forceps, scissors, needle drivers, clip appliers, graspers — all labeled class 0).

**Dataset split**

┌───────┬────────┐
│ Split │ Images │ Labels  │
├───────┼────────┤
│ train │ 22,731 │ 22,731   │
├───────┼────────┤
│ val   │ 3,536  │ 3,536      │
└───────┴────────┘

Composition: 10,326 real clinical images (CholecSeg8k, EndoVis2017/2018, SAR-RARP50, Roboflow), 1,732 real images from our own camera rig, 2,800 synthetic composites (real instrument crops pasted onto varied real/COCO backgrounds via feathered alpha blending + LAB color-matching), 7,873 negatives (COCO, mined confusors, clinical tissue-only frames).

Training arguments

epochs=30, batch=12, imgsz=1024, optimizer=AdamW
lr0=0.002, lrf=0.0002, cos_lr=True, warmup_epochs=1.0
degrees=180.0, flipud=0.5, fliplr=0.5, scale=0.3, translate=0.1, perspective=0.0005
hsv_h=0.0, hsv_s=0.2, hsv_v=0.3
mosaic=0.5, mixup=0.05, close_mosaic=5
resume=False

Final validation metrics (epoch 30)                                                           
box: 
precision: 0.8415
recall: 0.4725
mAP50: 0.5561
mAP50-95: 0.4108

mask
precision: 0.8484
recall: 0.441
mAP50: 0.4958
mAP50-95: 0.3166
                                 
The actual issue, concisely                                                                   
- Confusor false positives: non-surgical objects sharing superficial features with instruments(metal rods, screwdrivers) occas the 20-40% confidence range.
- Validation metrics vs. live performance disconnect: per-epoch confidence on 4 hand-verified live-camera regression frames sw from ~0.79 mean at epochs 7/10/12down to ~0.02-0.07 by epoch 20+ — while validation-set mAP/recall climbed steadily and monotonically the entire 30 epochs never reflecting this. best.pt landed in a low-live-performance epoch.

<img width="3000" height="2250" alt="confusion_matrix" src="https://github.com/user-attachments/assets/8433349f-86c4-4324-9efc-efc7eca2858b" />
<img width="3000" height="2250" alt="confusion_matrix_normalized" src="https://github.com/user-attachments/assets/e28f874e-2f4f-47eb-9bf6-1eccec6dd242" />
<img width="4000" height="1200" alt="results" src="https://github.com/user-attachments/assets/4fd2e45f-6f30-4535-8b67-20462423187d" />

**Comment by Kxrma47** [comment]
Those numbers point to validation-domain mismatch, not a need to force confidence to 0.8. Only 1,732 of 22,731 training images are camera-native, and live checkpoint ranking diverges from validation, so `best.pt` is optimizing the wrong distribution. Create a held-out camera-only validation set, save each epoch, and rank checkpoints on it:

```python
from pathlib import Path
from ultralytics import YOLO
for w in sorted(Path("runs/segment/train/weights").glob("epoch*.pt")):
    r = YOLO(w).val(data="live_camera.yaml", imgsz=1024, conf=0.001)
    print(w.name, r.seg.map50, r.seg.map)
```

Then oversample camera-native frames and hard-negative-mine the live false positives. Choose the inference threshold from that set's precision/recall curve; 70–80% is not a universal quality target. If SAM2 only refines masks, prompting it with YOLO boxes also avoids duplicating segmentation. References: [Ultralytics validation](https://docs.ultralytics.com/modes/val/) and [training settings](https://docs.ultralytics.com/modes/train/).

**Reply by pderrenger** [reply]
Treat this as a domain-generalization problem rather than targeting a fixed confidence score: validate on held-out frames from the actual live camera, add diverse camera-native positives, and iteratively retrain with real false positives as hard negatives. Calibrate `conf` from the camera validation precision-recall curve, and use YOLO11 detection boxes to prompt SAM 2 if SAM 2 is responsible for the final masks.

**Reply by robotricks16** [reply]
thanks for the reply i ran this snippet, across each epoch , turns out the there was a leakage in the custom negative dataset, so i decided to remove it, and trained again on yolo11s and yolo11n, 
2 models i trained on yolo11n:
when hue_h=0.0, i found out that the model is very confident when the background is reddish/pink, due to large dataset on surgeries videos, but on other bg, it losses conf on the surgical instruments 

<img width="3000" height="2250" alt="confusion_matrix_normalized" src="https://github.com/user-attachments/assets/4a6d9214-a164-4316-98c5-02844cf17db6" />
<img width="4000" height="1200" alt="results" src="https://github.com/user-attachments/assets/e7704f3d-b586-497d-b949-57de6e8d8467" />
<img width="3000" height="2250" alt="confusion_matrix" src="https://github.com/user-attachments/assets/c7b94d22-f920-459e-ac4d-692cc66250bb" />


- and when hue_h=0.5, model becomes background agnostic, but it was detecting teh elongated pointed objects like pen, needle also at around 0.80 conf

<img width="4000" height="1200" alt="results" src="https://github.com/user-attachments/assets/be706b8e-c5cc-46b2-aeb9-c42bcde4e4c2" />
<img width="3000" height="2250" alt="confusion_matrix_normalized" src="https://github.com/user-attachments/assets/f69a729e-996f-4d45-abc7-5eb911a9f974" />
<img width="3000" height="2250" alt="confusion_matrix" src="https://github.com/user-attachments/assets/b5225b72-0105-4022-98eb-99dbd2f6c3a0" />

**Comment by Kxrma47** [comment]
The leakage finding explains why the original validation looked better than deployment. The new result also makes sense: `hsv_h=0.5` is extremely strong (Ultralytics defaults to `0.015`), so it removes the pink-tissue shortcut but encourages the model to rely on shape; pens and needles then become high-confidence confusors.

I would keep the camera-only validation set fixed and run a one-variable sweep, for example:

```python
for h in (0.015, 0.03, 0.05, 0.10):
    model.train(data="surgical.yaml", hsv_h=h, name=f"hsv_h_{h}")
```

Compare camera-set precision/recall and false positives per image, not training mAP alone. Add the pen, needle, screwdriver, and metal-rod false positives as hard negatives captured through the same camera, under both tissue-like and neutral backgrounds. Also hold the architecture fixed while tuning hue so YOLO11n vs YOLO11s capacity does not confound the experiment.

Reference: [Ultralytics augmentation guide](https://docs.ultralytics.com/guides/yolo-data-augmentation/#hue-adjustment-hsv_h).

**Reply by robotricks16** [reply]
shall i negatively train on the non surgical (but elongated and pointed) tools like pens, markers, needles?
but there can be n number of datasets that can be long or elongated..

also my current dataset (on which i trained these models), consist of only positive images of surgeries of endovis, choleseg, and val dataset is also made from these datasets only

**Reply by robotricks16** [reply]
also, i trained a new model (yolo11n) in which i replaced global HSV jitter with a differential mask-aware transformation. Using ground-truth segmentation masks, i applied aggressive 360° hue augmentation to background tissue while restricting foreground surgical instruments to a tight ±3.6° realistic metallic hue range. i also applied shared saturation/value scaling for global photometric consistency., 

so i tested the best weights on the random non surgical instruments i still i got 0.86 confi on the live pen pic

from ultralytics import YOLO
model = YOLO("/file_dir/best.pt")
results = model("imagescrew.png")
results[0].show()
results[0].save(filename="prediction_result.jpg")

**Comment by Kxrma47** [comment]
Yes—add pens, needles, markers, screwdrivers, and similar objects as **background-only hard negatives**, but you do not need examples of every elongated object. Sample confusers from the actual deployment environment, train, run the model on new camera footage, collect its false positives, and repeat. That active hard-negative loop targets the mistakes the model really makes.

The larger problem is the current validation set: if both training and validation contain only EndoVis/CholeSeg-style surgical footage, validation cannot measure either non-surgical false positives or camera-domain shift. Build disjoint validation subsets:

- camera-native surgical positives;
- camera-native non-surgical confusers;
- tissue/background-only frames.

Split by video/session, not individual frames, so near-duplicate frames cannot cross train and validation. For a single `surgical_instrument` class, confuser images should have no objects labeled; do not create a “pen” class unless you actually need that classification output.

The mask-aware hue augmentation can reduce the tissue-color shortcut, but it cannot teach the semantic difference between two objects with similar geometry. The 0.86 pen prediction shows that shape is now the shortcut. Keep hue moderate, add real camera-native positives and hard negatives, and select the checkpoint/threshold using surgical recall together with false positives per negative image.

Dataset format: [Ultralytics segmentation datasets](https://docs.ultralytics.com/datasets/segment/). Validation guidance: [Ultralytics validation mode](https://docs.ultralytics.com/modes/val/).

**Reply by glenn-jocher** [reply]
Don’t optimize for a fixed 70–80% confidence—confidence is not an accuracy metric. Calibrate the threshold on camera-native validation data, then use YOLO11-seg to initialize or reacquire tools while [SAM 2 video propagation](https://docs.ultralytics.com/models/sam-2) maintains masks between detections; requiring temporal persistence across several frames before initializing SAM 2 can further suppress one-frame false positives.

**Reply by robotricks16** [reply]
"Sample confusers from the actual deployment environment, train, run the model on new camera footage, collect its false positives, and repeat. That active hard-negative loop targets the mistakes the model really makes."

Ok i already have witnessed to train the model on camera native negative, like screw, needle, screw driver, etc.
(The above two models that i have shared were trained camera native negative like screwdriver, cylindrical rod, needle, scissors, etc and also artnet negative dataset of tissues images only, results i have shared).

But this time a new camera native i found was pen and marker, on both of them i got 0.7 to 0.85 conf, my point is , is this valid practise to repeatedly collect negative camera native images and repeat this, to train the model on various elongated object with pointed end, as there can be n number of objects in the world, its like hardcoding the model, to ignore them. I can create a val folder with camera native positive that i have, but not on all objects that are elongated and pointed end, that resembles the surgical instruments. So training all of them negatively wont be possible i guess.

I hope you will understand my confusion and my concern. Pls let me know how this particular issue can be solved

**Comment by Kxrma47** [comment]
Your concern is valid, but iterative hard-negative mining is standard practice rather than hardcoding individual objects. The goal is not to enumerate every object; it is to sample the main **confuser families** the deployment camera produces—for example thin plastic cylinders, shiny metal rods, hinged/scissor-like objects, and pointed objects—so the model learns a broader boundary.

Use a fixed experiment:

1. Split by camera session/video.
2. Put some confuser families in training, but reserve entire families and sessions as unseen validation/test negatives.
3. After each run, add false positives only to the training pool; never move the held-out test examples into training.
4. Measure recall on camera-native surgical tools and false positives per negative image.

If performance improves on unseen confuser families, mining is generalizing. If it only remembers the mined objects, the fixed held-out score will expose that. Also keep adding diverse camera-native **positive** tools; adding only negatives can reduce recall.

There is a fundamental limit: if a pen and a surgical instrument are visually indistinguishable at the available resolution, a single-frame binary detector has no evidence from which to infer their intended use. Then the solution must add information—higher-resolution/part-level features, a larger model tested on the same split, a restricted surgical-workspace ROI, or temporal/context rules. The temporal persistence suggested by @glenn-jocher can suppress brief false detections, but a persistent pen may still pass; SAM 2 propagates the prompted mask and does not determine whether the object is surgical.

Finally, `0.85` confidence is not an 85% probability that the object is a surgical tool. Calibrate the operating threshold on the fixed camera-native validation set and report both surgical recall and negative-set false positives. References: [Ultralytics validation](https://docs.ultralytics.com/modes/val/) and [tracking mode](https://docs.ultralytics.com/modes/track/).

**Reply by glenn-jocher** [reply]
The key is to optimize camera-native validation precision/recall rather than target `0.7–0.8` confidence: add diverse in-domain positives, verify tight/consistent masks, split by video/session, test larger `imgsz` and model sizes, and mine false positives only from the training split. If YOLO11-seg masks are sufficient, use them directly; otherwise use YOLO prompts to initialize [SAM 2 video segmentation](https://docs.ultralytics.com/models/sam-2) and propagate masks temporally.

**Comment by navalkishoremehta95** [comment]
A few practical notes from similar detector → SAM prompting setups:

**On false positives (pens / scissors / rods)**  
Negative images help, but if those objects still look like tools under metal glare, YOLO will keep firing. Better pattern:
1. Keep one class (or a small set) for *surgical tool*.
2. Add hard negatives as **background** (no labels), not as extra classes, unless you really need to classify them.
3. Prefer **more in-domain surgical frames** (different lighting, smoke, blood, specular metal) over more COCO clutter. Domain shift is usually the main confidence killer.

**On low confidence (20–50%)**  
Expect that when train data is mostly red tissue and the live feed is different. Confidence thresholds are not a substitute for matching the camera domain. Try:
- train `yolo11s-seg` longer with a lower LR after a short freeze of the backbone
- stronger color jitter / mosaic off for late epochs (`close_mosaic`)
- evaluate with the **same resolution and color space** as the OR camera
- calibrate on a held-out OR clip; set the SAM trigger from precision–recall there, not from a fixed 0.7

**Pipeline tip**  
For live feed, use YOLO boxes (or masks) only as SAM2 prompts, then **filter by geometry** (aspect ratio, length, metal-like appearance) before accepting a mask. That often beats chasing 80% YOLO conf alone.

If you can share a few false-positive and low-conf frames (and whether train/val are from the same camera), it is easier to say whether you need more data or just thresholding / NMS tuning.

**Reply by pderrenger** [reply]
Don’t optimize for an arbitrary `0.7–0.8` confidence; build a held-out validation set from the actual live-feed camera and select the trigger threshold from the mask PR/F1 curves using [Ultralytics validation](https://docs.ultralytics.com/modes/val). Keep confusing objects as unlabeled hard-negative images, prioritize production-camera frames across lighting, glare, smoke, occlusion, and backgrounds, and split by video/scene to prevent adjacent-frame leakage. SAM 2 can then use the accepted YOLO prompt and its video memory for subsequent frames; please share the `MaskPR_curve.png`, `MaskF1_curve.png`, dataset counts, and representative false-positive/low-confidence frames for more targeted guidance.
